opm/input/eclipse/Schedule/Action/ActionParser.cpp

```diff
@@ -17,324 +17,583 @@
+{
+    if (this->parseIsComplete()) {
+        throw std::logic_error {
+            "Parse process proceeds beyond token sequence end"
+        };
     }
 
-    return TokenType::ecl_expr;
+    return this->tokens_[this->current_pos_];
 }
 
-FuncType Parser::get_func(const std::string& arg)
+Opm::Action::TokenType ActionParser::currentType() const
 {
-    if (arg == "YEAR") return FuncType::time;
-    if (arg == "MNTH") return FuncType::time_month;
-    if (arg == "DAY")  return FuncType::time;
-
-    using Cat = SummaryConfigNode::Category;
-    SummaryConfigNode::Category cat = parseKeywordCategory(arg);
-    switch (cat) {
-        case Cat::Aquifer:    return FuncType::aquifer;
-        case Cat::Well:       return FuncType::well;
-        case Cat::Group:      return FuncType::group;
-        case Cat::Connection: return FuncType::well_connection;
-        case Cat::Region:     return FuncType::region;
-        case Cat::Block:      return FuncType::block;
-        case Cat::Segment:    return FuncType::well_segment;
-        default:              return FuncType::none;
-    }
+    return this->parseIsComplete()
+        ? Opm::Action::TokenType::end
+        : Opm::Action::Parser::tokenType(this->currentToken());
 }
 
-ParseNode Parser::next()
+ActionParser::Node ActionParser::current() const
 {
-    this->current_pos++;
-    if (static_cast<size_t>(this->current_pos) == this->tokens.size()) {
-        return ParseNode(TokenType::end);
+    if (this->parseIsComplete()) {
+        return Node { Opm::Action::TokenType::end };
     }
 
-    std::string arg = this->tokens[this->current_pos];
-    return ParseNode(get_type(arg), arg);
+    const auto& arg = this->currentToken();
+    return Node { Opm::Action::Parser::tokenType(arg), arg };
 }
 
-ParseNode Parser::current() const
+bool ActionParser::parseIsComplete() const
 {
-    if (static_cast<size_t>(this->current_pos) == this->tokens.size()) {
-        return ParseNode(TokenType::end);
-    }
+    return static_cast<TokenSize>(this->current_pos_)
+        >= this->tokens_.size();
+}
+
+void ActionParser::advanceTokenPosition()
+{
+    ++this->current_pos_;
+}
 
-    std::string arg = this->tokens[this->current_pos];
-    return ParseNode(get_type(arg), arg);
+ActionParser::Node ActionParser::next()
+{
+    this->advanceTokenPosition();
+    return this->current();
 }
 
-ASTNode Parser::parse_left()
+Opm::Action::ASTNode ActionParser::parse_left()
 {
     auto curr = this->current();
-    if (curr.type != TokenType::ecl_expr) {
+    if (curr.type != Opm::Action::TokenType::ecl_expr) {
         throw std::invalid_argument {
             fmt::format("Expected expression as left hand side "
                         "of comparison, but got {} instead.",
                         curr.value)
         };
     }
 
-    std::string func = curr.value;
-    FuncType func_type = get_func(curr.value);
-    std::vector<std::string> arg_list;
+    const auto func = curr.value;
```

Comment:
can be const ref.

Reply:
> can be const ref.

I don't think so.  We [update `curr`](https://github.com/OPM/opm-common/blob/bb8562aad5ff01521250380c8ecd432df3fe7519/opm/input/eclipse/Schedule/Action/ActionParser.cpp#L379) in the following loop, whence this might either be a dangling reference or it might refer to some other object.